rfcs/0000-safe-transmute.md

```diff
@@ -0,0 +1,2150 @@
+
+The question is, then: *how can the author of a type reason about transmutations they did not write, from-or-to types they did not write?* We address this problem by introducing two traits which both allow an author to opt-in to stability guarantees for their types, and allow third-parties to reason at compile-time about what guarantees are provided for such types.
+
+#### `PromiseTransmutableFrom` and `PromiseTransmutableInto`
+
+You may declare the stability guarantees of your type by implementing one or both of two traits:
+```rust
+pub trait PromiseTransmutableFrom
+{
+    type Archetype
+        : TransmuteInto<Self, NeglectStability>
+        + PromiseTransmutableFrom;
+}
+
+pub trait PromiseTransmutableInto
+{
+    type Archetype
+        : TransmuteFrom<Self, NeglectStability>
+        + PromiseTransmutableInto;
+}
+```
+
+To implement each of these traits, you must specify an `Archetype`. An `Archetype` is a type whose layout exemplifies the extremities of your stability promise (i.e., the least/most constrained type for which it is valid to transmute your type into/from).
+
+By implementing `PromiseTransmutableFrom`, you promise that your type is guaranteed to be safely transmutable *from* `PromiseTransmutableFrom::Archetype`. Conversely, by implementing `PromiseTransmutableInto`, you promise that your type is guaranteed to be safely transmutable *into* `PromiseTransmutableInto::Archetype`.
+
+You are free to change the layout of your type however you like between minor crate versions so long as that change does not violates these promises. These two traits are capable of expressing simple and complex stability guarantees.
+
+#### Stability & Transmutation
+A `Src` type is *stably* transmutable into a `Dst` type *only if* `<Src as PromiseTransmutableInto>::Archetype` is transmutable, stability notwithstanding, into `<Dst as PromiseTransmutableFrom>::Archetype`; formally:
+```rust
+unsafe impl<Src, Dst> TransmuteFrom<Src> for Dst
+where
+    Src: PromiseTransmutableInto,
+    Dst: PromiseTransmutableFrom,
+    <Dst as PromiseTransmutableFrom>::Archetype:
+        TransmuteFrom<
+            <Src as PromiseTransmutableInto>::Archetype,
+            NeglectStability
+        >
+{}
+```
+
+Why is this the case?
+
+The type `<Src as PromiseTransmutableInto>::Archetype` exemplifies the furthest extreme of non-breaking changes that could be made to the layout of `Src` that could affect its use as a source type in transmutations. Conversely, `<Dst as PromiseTransmutableFrom>::Archetype` exemplifies the furthest extreme of non-breaking changes that could be made to the layout of `Dst` that could affect its use as a destination type in transmutations. If a transmutation between these extremities is valid, then so is `Src: TransmuteInto<Dst>`.
+
+#### Common Use-Case: Promising a Fixed Layout
```

Comment:
@rylev I'm unsure how useful this shorthand would be.

For end-users, I anticipate that `PromiseTransmutableFrom` and `PromiseTransmutableInto` will virtually never appear in users' code outside of the context of implementing those traits. In writing the examples for this RFC, I haven't yet encountered an instance where I needed to use either of those traits in a `where` bound. So, I don't think this shorthand would be useful outside of the context of implementing the traits.

And, _within_ the context of implementing those traits, nearly all users will be using the shorthand `#[derive(PromiseTransmutableFrom, PromiseTransmutableInto)]`. So here, `#[derive(FixedLayout)]` would be a shorthand for a shorthand.

Is there a use-case you have in mind?

Reply:
Short-hand for short-hand is not necessarily a bad thing 😄 . I believe a vast majority of the time that users will want `#[derive(PromiseTransmutableFrom, PromiseTransmutableInto)]`. Given the long names and the frequency of its use `#[derive(FixedLayout)]` strikes me as an ergonomic choice. But I realize that it's not super elegant to introduce a new trait just as a short-hand for deriving.

Reply:
No, that's a better trait name for sure. People will understand it more when looking at docs and source.

Reply:
@rylev  Defining a trait is rather tricky (since those derives are a little more nuanced than just `Archetype = Self` to account for field stability declarations), but it's very technically feasible to provide a `derive(PromiseTransmutableFromAndInto)` macro that would be equivalent to `derive(PromiseTransmutableFrom, PromiseTransmutableInto)` — *without* a `PromiseTransmutableFromAndInto` trait.

Reply:
> Defining a trait is rather tricky (since those derives are a little more nuanced than just Archetype = Self to account for field stability declarations)

To clarify a little more about why having a `FixedLayout` shorthand/trait is tricky, it's because the concept of "fixed layout" is a little tricky.

---

First, there's the name: `PromiseTransmutableFrom` isn't *really* making a promise about layout, it's making a promise about future transmutability. Transmutability involves factors beyond layout, like [constructability](https://github.com/jswrenn/project-safe-transmute/blob/rfc/rfcs/0000-safe-transmute.md#-when-is-a-transmutation-safe). If we wanted to create a `derive` shorthand for `derive(PromiseTransmutableFrom, PromiseTransmutableInto)`, the name `PromiseFixedLayout` (or `PromiseStableLayout`, etc...) misrepresents what's actually being promised.

---

Second, there's the semantics: what does promising a fixed layout *mean*? You're probably mean to promise that *you* won't make changes to the type's transmutability, but what about *others*? What if your type's fields come from other crates? What if your type's fields are generic?

The behavior of `#[derive(PromiseTransmutableFrom, PromiseTransmutableInto)]` reflects this inherent complexity: it **doesn't** mean your type has *totally-assured-transmutability*, it means your type has *as-assured-as-possible-transmutability*. Concretely: When you write `#[derive(PromiseTransmutable{From,Into})]`, your type [inherits the instability of its fields](https://github.com/jswrenn/project-safe-transmute/blob/rfc/rfcs/0000-safe-transmute.md#common-use-case-promising-a-fixed-layout). If its fields have strong stability promises, so too will your type; if its fields have weak stability promises, so too will your type.

---

To conclude, I don't think there's a good mapping of a [mem-markers](https://github.com/rust-secure-code/mem-markers)-style `FixedLayout` abstraction onto  `PromiseTransmutableFrom` and `PromiseTransmutableInto`. While it might seem simpler, at first glance, to just present *one* trait (`FixedLayout`) instead of two (`PromiseTransmutableFrom` and `PromiseTransmutableInto`), the concept of a `FixedLayout` isn't well-defined, and the complexities of `PromiseTransmutable{From,Into}` reflect the complexity *inherent* to the problem of transmutation stability.

Reply:
What I'm arguing here is that there is a use case for "POD" that will be quite common. By "POD" in this case I mean types where the in memory-representation of that type is intrinsic to its functionality and changing that would be a breaking change. This is common in the parsing scenario for instance. Having some sort of short hand for saying this type's layout/alignment/etc cannot change would be a simple "happy path" for many use cases. I'm not arguing this as a replacement to `PromiseTransmutable{From,Into}` but as a compliment. This doesn't have to be a `derive` either. I can imagine also `#[repr(C, fixed)]` for instance.

Reply:
Per a conversation with @rylev, https://github.com/rust-lang/project-safe-transmute/pull/5/commits/ed172fe84874c154118f2054435ceb8ff94312e1 proposes a `#[derive(PromiseTransmutable)]` ergonomic extension. I'm fairly [convinced](https://github.com/jswrenn/project-safe-transmute/blob/rfc/rfcs/0000-safe-transmute.md#extension-promisetransmutable-shorthand) that adding such a shorthand is the right thing to do and I [refer to it in the Guide-Level Explanation](https://github.com/jswrenn/project-safe-transmute/blob/rfc/rfcs/0000-safe-transmute.md#common-use-case-as-stable-as-possible), but it's formally defined as an Extension because it's technically non-central to the RFC and poses a few unusual design quirks.